File: src/python/jvdb/jvdb2.py
"""
JVDB 2.0 - Clase de Abstracción para Acceso a Base de Datos
Versión: 2.0.0

Versión mejorada con:
- Sistema de logging integrado
- Pool de conexiones
- Soporte para transacciones
- CRUD completo optimizado
- Manejo robusto de errores
"""

import json
from typing import Optional, List, Dict, Any, Union
from contextlib import contextmanager

try:
    # Importaciones relativas (cuando se usa como paquete)
    from ..yoursql import YourSQLConnection
    from .connection_pool import JVConnectionPool, PooledConnection
    from .logger import JVLogger
    from .exceptions import *
except ImportError:
    # Importaciones absolutas (cuando se ejecuta directamente)
    import sys
    import os
    sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
    from yoursql.yoursql import YourSQLConnection
    from jvdb.connection_pool import JVConnectionPool, PooledConnection
    from jvdb.logger import JVLogger
    from jvdb.exceptions import *


class JVDB2:
    """
    Clase de abstracción mejorada para acceso a base de datos MySQL
    Versión 2.0 con pool de conexiones, logging y transacciones
    """
    
    def __init__(
        self,
        host: str,
        usuario: str,
        contrasena: str,
        basedatos: str,
        port: int = 3306,
        use_pool: bool = True,
        pool_config: Optional[Dict] = None,
        log_config: Optional[Dict] = None
    ):
        """
        Inicializa JVDB2 con los parámetros de conexión
        
        Args:
            host: Dirección del servidor MySQL
            usuario: Usuario de la base de datos
            contrasena: Contraseña del usuario
            basedatos: Nombre de la base de datos
            port: Puerto de conexión (por defecto 3306)
            use_pool: Usar pool de conexiones (por defecto True)
            pool_config: Configuración del pool
            log_config: Configuración del logger
        """
        self.host = host
        self.usuario = usuario
        self.contrasena = contrasena
        self.basedatos = basedatos
        self.port = port
        
        # Configurar logger
        log_cfg = log_config or {}
        self.logger = JVLogger.get_instance("JVDB2", **log_cfg)
        
        # Pool de conexiones o conexión simple
        self.use_pool = use_pool
        self.pool = None
        self.conexion = None
        
        if use_pool:
            pool_cfg = pool_config or {}
            self.pool = JVConnectionPool(
                host=host,
                user=usuario,
                password=contrasena,
                database=basedatos,
                port=port,
                min_size=pool_cfg.get('min_size', 2),
                max_size=pool_cfg.get('max_size', 10),
                max_lifetime=pool_cfg.get('max_lifetime', 3600),
                timeout=pool_cfg.get('timeout', 30),
                logger=self.logger
            )
            self.logger.info(f"JVDB2 inicializado con pool de conexiones")
        else:
            self.conexion = YourSQLConnection(
                host=host,
                user=usuario,
                password=contrasena,
                database=basedatos,
                port=port
            )
            if self.conexion.connect():
                self.logger.connection(host, basedatos, True)
            else:
                self.logger.connection(host, basedatos, False)
                raise ConnectionError(f"No se pudo conectar a {host}/{basedatos}")
    
    @contextmanager
    def _get_connection(self):
        """Context manager para obtener una conexión"""
        if self.use_pool:
            pooled_conn = self.pool.acquire()
            try:
                yield pooled_conn.connection
            finally:
                self.pool.release(pooled_conn)
        else:
            yield self.conexion
    
    def seleccionar(
        self,
        tabla: str,
        columnas: Union[str, List[str]] = "*",
        where: Optional[Dict[str, Any]] = None,
        order_by: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
        formato: str = "list"
    ) -> Union[str, List[Dict]]:
        """
        Selecciona registros de una tabla con opciones avanzadas
        
        Args:
            tabla: Nombre de la tabla
            columnas: Columnas a seleccionar (lista o string)
            where: Condiciones WHERE como diccionario
            order_by: Ordenamiento (ej: "columna DESC")
            limit: Límite de resultados
            offset: Offset para paginación
            formato: 'json' o 'list' (por defecto list)
            
        Returns:
            Datos en el formato especificado
        """
        try:
            # Construir columnas
            if isinstance(columnas, list):
                cols = ", ".join(columnas)
            else:
                cols = columnas
            
            # Construir query
            query = f"SELECT {cols} FROM {tabla}"
            params = []
            
            # Agregar WHERE
            if where:
                conditions = []
                for col, val in where.items():
                    conditions.append(f"{col} = %s")
                    params.append(val)
                query += " WHERE " + " AND ".join(conditions)
            
            # Agregar ORDER BY
            if order_by:
                query += f" ORDER BY {order_by}"
            
            # Agregar LIMIT y OFFSET
            if limit:
                query += f" LIMIT {limit}"
                if offset:
                    query += f" OFFSET {offset}"
            
            self.logger.query(query, tuple(params) if params else None)
            
            with self._get_connection() as conn:
                resultado = conn.execute_query(
                    query,
                    params=tuple(params) if params else None,
                    fetch_all=True
                )
            
            if resultado is None:
                resultado = []
            
            self.logger.info(f"Seleccionados {len(resultado)} registros de '{tabla}'")
            
            if formato == "json":
                return json.dumps(resultado, ensure_ascii=False, indent=2)
            return resultado
            
        except Exception as e:
            self.logger.error(f"Error en seleccionar: {e}", exc_info=True)
            raise QueryError(f"Error al seleccionar de {tabla}: {e}")
    
    def seleccionar_uno(
        self,
        tabla: str,
        identificador: Optional[int] = None,
        where: Optional[Dict[str, Any]] = None
    ) -> Optional[Dict]:
        """
        Selecciona un único registro
        
        Args:
            tabla: Nombre de la tabla
            identificador: ID del registro (campo Identificador)
            where: Condiciones alternativas como diccionario
            
        Returns:
            Diccionario con el registro o None
        """
        try:
            if identificador is not None:
                where = {'Identificador': identificador}
            elif where is None:
                raise ValidationError("Debe proporcionar identificador o where")
            
            resultados = self.seleccionar(tabla, where=where, limit=1, formato="list")
            return resultados[0] if resultados else None
            
        except Exception as e:
            self.logger.error(f"Error en seleccionar_uno: {e}", exc_info=True)
            raise QueryError(f"Error al seleccionar uno de {tabla}: {e}")
    
    def insertar(
        self,
        tabla: str,
        datos: Dict[str, Any],
        return_id: bool = False
    ) -> Union[int, Dict[str, Any]]:
        """
        Inserta un nuevo registro
        
        Args:
            tabla: Nombre de la tabla
            datos: Diccionario con columna:valor
            return_id: Si True, devuelve dict con rows_affected y last_id
            
        Returns:
            Número de filas afectadas o dict con info
        """
        try:
            if not datos:
                raise ValidationError("No se proporcionaron datos para insertar")
            
            columnas = ", ".join(datos.keys())
            placeholders = ", ".join(["%s"] * len(datos))
            valores = tuple(datos.values())
            
            query = f"INSERT INTO {tabla} ({columnas}) VALUES ({placeholders})"
            self.logger.query(query, valores)
            
            with self._get_connection() as conn:
                resultado = conn.execute_query(query, params=valores)
                last_id = conn._connection.get_server_info() if hasattr(conn, '_connection') else None
            
            self.logger.info(f"Registro insertado en '{tabla}'")
            
            if return_id:
                return {
                    'rows_affected': resultado if resultado else 0,
                    'last_insert_id': last_id
                }
            return resultado if resultado else 0
            
        except Exception as e:
            self.logger.error(f"Error en insertar: {e}", exc_info=True)
            raise QueryError(f"Error al insertar en {tabla}: {e}")
    
    def insertar_multiple(
        self,
        tabla: str,
        registros: List[Dict[str, Any]]
    ) -> int:
        """
        Inserta múltiples registros de forma optimizada
        
        Args:
            tabla: Nombre de la tabla
            registros: Lista de diccionarios con los datos
            
        Returns:
            Número total de filas insertadas
        """
        try:
            if not registros:
                return 0
            
            total = 0
            with self.transaction():
                for registro in registros:
                    total += self.insertar(tabla, registro)
            
            self.logger.info(f"{total} registros insertados en '{tabla}'")
            return total
            
        except Exception as e:
            self.logger.error(f"Error en insertar_multiple: {e}", exc_info=True)
            raise QueryError(f"Error al insertar múltiples en {tabla}: {e}")
    
    def actualizar(
        self,
        tabla: str,
        datos: Dict[str, Any],
        identificador: Optional[int] = None,
        where: Optional[Dict[str, Any]] = None
    ) -> int:
        """
        Actualiza registros existentes
        
        Args:
            tabla: Nombre de la tabla
            datos: Diccionario con columna:valor a actualizar
            identificador: ID del registro (campo Identificador)
            where: Condiciones alternativas como diccionario
            
        Returns:
            Número de filas afectadas
        """
        try:
            if not datos:
                raise ValidationError("No se proporcionaron datos para actualizar")
            
            if identificador is not None:
                where = {'Identificador': identificador}
            elif where is None:
                raise ValidationError("Debe proporcionar identificador o where")
            
            # Construir SET clause
            set_clause = ", ".join([f"{col} = %s" for col in datos.keys()])
            valores = list(datos.values())
            
            # Construir WHERE clause
            where_conditions = []
            for col, val in where.items():
                where_conditions.append(f"{col} = %s")
                valores.append(val)
            
            query = f"UPDATE {tabla} SET {set_clause} WHERE " + " AND ".join(where_conditions)
            self.logger.query(query, tuple(valores))
            
            with self._get_connection() as conn:
                resultado = conn.execute_query(query, params=tuple(valores))
            
            self.logger.info(f"{resultado} registro(s) actualizado(s) en '{tabla}'")
            return resultado if resultado else 0
            
        except Exception as e:
            self.logger.error(f"Error en actualizar: {e}", exc_info=True)
            raise QueryError(f"Error al actualizar {tabla}: {e}")
    
    def eliminar(
        self,
        tabla: str,
        identificador: Optional[int] = None,
        where: Optional[Dict[str, Any]] = None
    ) -> int:
        """
        Elimina registros de una tabla
        
        Args:
            tabla: Nombre de la tabla
            identificador: ID del registro (campo Identificador)
            where: Condiciones alternativas como diccionario
            
        Returns:
            Número de filas eliminadas
        """
        try:
            if identificador is not None:
                where = {'Identificador': identificador}
            elif where is None:
                raise ValidationError("Debe proporcionar identificador o where")
            
            # Construir WHERE clause
            where_conditions = []
            valores = []
            for col, val in where.items():
                where_conditions.append(f"{col} = %s")
                valores.append(val)
            
            query = f"DELETE FROM {tabla} WHERE " + " AND ".join(where_conditions)
            self.logger.query(query, tuple(valores))
            
            with self._get_connection() as conn:
                resultado = conn.execute_query(query, params=tuple(valores))
            
            self.logger.info(f"{resultado} registro(s) eliminado(s) de '{tabla}'")
            return resultado if resultado else 0
            
        except Exception as e:
            self.logger.error(f"Error en eliminar: {e}", exc_info=True)
            raise QueryError(f"Error al eliminar de {tabla}: {e}")
    
    def contar(
        self,
        tabla: str,
        where: Optional[Dict[str, Any]] = None
    ) -> int:
        """
        Cuenta registros en una tabla
        
        Args:
            tabla: Nombre de la tabla
            where: Condiciones opcionales
            
        Returns:
            Número de registros
        """
        try:
            query = f"SELECT COUNT(*) as total FROM {tabla}"
            params = []
            
            if where:
                conditions = []
                for col, val in where.items():
                    conditions.append(f"{col} = %s")
                    params.append(val)
                query += " WHERE " + " AND ".join(conditions)
            
            with self._get_connection() as conn:
                resultado = conn.execute_query(
                    query,
                    params=tuple(params) if params else None,
                    fetch_one=True
                )
            
            return resultado['total'] if resultado else 0
            
        except Exception as e:
            self.logger.error(f"Error en contar: {e}", exc_info=True)
            raise QueryError(f"Error al contar en {tabla}: {e}")
    
    def existe(
        self,
        tabla: str,
        where: Dict[str, Any]
    ) -> bool:
        """
        Verifica si existe al menos un registro que cumpla las condiciones
        
        Args:
            tabla: Nombre de la tabla
            where: Condiciones de búsqueda
            
        Returns:
            True si existe al menos un registro
        """
        return self.contar(tabla, where) > 0
    
    @contextmanager
    def transaction(self):
        """
        Context manager para transacciones
        
        Uso:
            with db.transaction():
                db.insertar('tabla', datos1)
                db.actualizar('tabla', datos2, id=1)
        """
        if not self.use_pool:
            conn = self.conexion
        else:
            pooled_conn = self.pool.acquire()
            conn = pooled_conn.connection
        
        try:
            # Iniciar transacción
            conn._connection.autocommit = False
            self.logger.transaction("BEGIN")
            
            yield conn
            
            # Commit
            conn._connection.commit()
            self.logger.transaction("COMMIT")
            
        except Exception as e:
            # Rollback
            conn._connection.rollback()
            self.logger.transaction("ROLLBACK")
            self.logger.error(f"Error en transacción: {e}", exc_info=True)
            raise TransactionError(f"Error en transacción: {e}")
        
        finally:
            conn._connection.autocommit = True
            if self.use_pool:
                self.pool.release(pooled_conn)
    
    def consulta_personalizada(
        self,
        query: str,
        params: Optional[tuple] = None,
        fetch: str = "all"
    ) -> Any:
        """
        Ejecuta una consulta SQL personalizada
        
        Args:
            query: Consulta SQL
            params: Parámetros para consultas preparadas
            fetch: 'all', 'one', o 'none'
            
        Returns:
            Resultados según el tipo de consulta
        """
        try:
            self.logger.query(query, params)
            
            with self._get_connection() as conn:
                if fetch == "one":
                    resultado = conn.execute_query(query, params=params, fetch_one=True)
                elif fetch == "all":
                    resultado = conn.execute_query(query, params=params, fetch_all=True)
                else:
                    resultado = conn.execute_query(query, params=params)
            
            return resultado
            
        except Exception as e:
            self.logger.error(f"Error en consulta personalizada: {e}", exc_info=True)
            raise QueryError(f"Error en consulta: {e}")
    
    def get_pool_stats(self) -> Optional[Dict[str, Any]]:
        """Obtiene estadísticas del pool de conexiones"""
        if self.pool:
            return self.pool.get_stats()
        return None
    
    def cerrar(self):
        """Cierra el pool o la conexión"""
        try:
            if self.pool:
                self.pool.close()
                self.logger.info("Pool de conexiones cerrado")
            elif self.conexion:
                self.conexion.disconnect()
                self.logger.info("Conexión cerrada")
        except Exception as e:
            self.logger.error(f"Error al cerrar: {e}")
    
    def __enter__(self):
        """Soporte para context manager"""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Cierra al salir del contexto"""
        self.cerrar()
    
    def __del__(self):
        """Destructor"""
        try:
            self.cerrar()
        except:
            pass

File: src/php/JVDB/JVDB2.php
<?php
/**
 * JVDB 2.0 - Clase de Abstracción para Acceso a Base de Datos (PHP)
 * Versión: 2.0.0
 * 
 * Versión mejorada con:
 * - Sistema de logging integrado
 * - Pool de conexiones
 * - Soporte para transacciones
 * - CRUD completo optimizado
 * - Manejo robusto de errores
 */

namespace JVDB;

require_once __DIR__ . '/../YourSQL/YourSQL.php';
require_once __DIR__ . '/JVConnectionPool.php';
require_once __DIR__ . '/JVLogger.php';
require_once __DIR__ . '/JVExceptions.php';

use YourSQL\YourSQLConnection;
use JVDB\Exceptions\ValidationException;
use JVDB\Exceptions\QueryException;
use JVDB\Exceptions\ConnectionException;
use JVDB\Exceptions\TransactionException;
use JVDB\Exceptions\PoolExhaustedException;
use JVDB\Exceptions\ConfigurationException;

class JVDB2 {
    private string $host;
    private string $usuario;
    private string $contrasena;
    private string $basedatos;
    private int $port;
    private JVLogger $logger;
    private bool $usePool;
    private ?JVConnectionPool $pool = null;
    private ?YourSQLConnection $conexion = null;
    private bool $inTransaction = false;
    
    /**
     * Constructor - Inicializa JVDB2 con los parámetros de conexión
     * 
     * @param string $host Dirección del servidor MySQL
     * @param string $usuario Usuario de la base de datos
     * @param string $contrasena Contraseña del usuario
     * @param string $basedatos Nombre de la base de datos
     * @param int $port Puerto de conexión
     * @param bool $usePool Usar pool de conexiones
     * @param array $poolConfig Configuración del pool
     * @param array $logConfig Configuración del logger
     */
    public function __construct(
        string $host,
        string $usuario,
        string $contrasena,
        string $basedatos,
        int $port = 3306,
        bool $usePool = true,
        array $poolConfig = [],
        array $logConfig = []
    ) {
        $this->host = $host;
        $this->usuario = $usuario;
        $this->contrasena = $contrasena;
        $this->basedatos = $basedatos;
        $this->port = $port;
        $this->usePool = $usePool;
        
        // Configurar logger
        $this->logger = JVLogger::getInstance('JVDB2', $logConfig);
        
        // Pool de conexiones o conexión simple
        if ($usePool) {
            $this->pool = new JVConnectionPool(
                $host,
                $usuario,
                $contrasena,
                $basedatos,
                $port,
                $poolConfig['min_size'] ?? 2,
                $poolConfig['max_size'] ?? 10,
                $poolConfig['max_lifetime'] ?? 3600,
                $poolConfig['timeout'] ?? 30,
                $this->logger
            );
            $this->logger->info("JVDB2 inicializado con pool de conexiones");
        } else {
            $this->conexion = new YourSQLConnection($host, $usuario, $contrasena, $basedatos, $port);
            if ($this->conexion->connect()) {
                $this->logger->connection($host, $basedatos, true);
            } else {
                $this->logger->connection($host, $basedatos, false);
                throw new ConnectionException("No se pudo conectar a $host/$basedatos");
            }
        }
    }
    
    /**
     * Obtiene una conexión del pool o la conexión directa
     * 
     * @return array [PooledConnection|YourSQLConnection, bool shouldRelease]
     */
    private function getConnection(): array {
        if ($this->usePool && !$this->inTransaction) {
            $pooledConn = $this->pool->acquire();
            return [$pooledConn->connection, true, $pooledConn];
        } else {
            return [$this->conexion, false, null];
        }
    }
    
    /**
     * Libera una conexión si es necesario
     */
    private function releaseConnection($pooledConn): void {
        if ($pooledConn && $this->usePool && !$this->inTransaction) {
            $this->pool->release($pooledConn);
        }
    }
    
    /**
     * Selecciona registros de una tabla con opciones avanzadas
     * 
     * @param string $tabla Nombre de la tabla
     * @param string|array $columnas Columnas a seleccionar
     * @param array|null $where Condiciones WHERE
     * @param string|null $orderBy Ordenamiento
     * @param int|null $limit Límite de resultados
     * @param int|null $offset Offset para paginación
     * @param string $formato 'json' o 'array'
     * @return string|array Datos en el formato especificado
     */
    public function seleccionar(
        string $tabla,
        $columnas = '*',
        ?array $where = null,
        ?string $orderBy = null,
        ?int $limit = null,
        ?int $offset = null,
        string $formato = 'array'
    ) {
        try {
            // Construir columnas
            $cols = is_array($columnas) ? implode(', ', $columnas) : $columnas;
            
            // Construir query
            $query = "SELECT $cols FROM $tabla";
            $params = [];
            
            // Agregar WHERE
            if ($where) {
                $conditions = [];
                foreach ($where as $col => $val) {
                    $conditions[] = "$col = ?";
                    $params[] = $val;
                }
                $query .= " WHERE " . implode(' AND ', $conditions);
            }
            
            // Agregar ORDER BY
            if ($orderBy) {
                $query .= " ORDER BY $orderBy";
            }
            
            // Agregar LIMIT y OFFSET
            if ($limit) {
                $query .= " LIMIT $limit";
                if ($offset) {
                    $query .= " OFFSET $offset";
                }
            }
            
            $this->logger->query($query, $params);
            
            list($conn, $shouldRelease, $pooledConn) = $this->getConnection();
            $resultado = $conn->executeQuery($query, $params);
            $this->releaseConnection($pooledConn);
            
            if ($resultado === null) {
                $resultado = [];
            }
            
            $this->logger->info(count($resultado) . " registros seleccionados de '$tabla'");
            
            if ($formato === 'json') {
                return json_encode($resultado, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT);
            }
            
            return $resultado;
            
        } catch (\Exception $e) {
            $this->logger->error("Error en seleccionar: {$e->getMessage()}");
            throw new QueryException("Error al seleccionar de $tabla: {$e->getMessage()}");
        }
    }
    
    /**
     * Selecciona un único registro
     * 
     * @param string $tabla Nombre de la tabla
     * @param int|null $identificador ID del registro
     * @param array|null $where Condiciones alternativas
     * @return array|null Registro o null
     */
    public function seleccionarUno(string $tabla, ?int $identificador = null, ?array $where = null): ?array {
        try {
            if ($identificador !== null) {
                $where = ['Identificador' => $identificador];
            } elseif ($where === null) {
                throw new ValidationException("Debe proporcionar identificador o where");
            }
            
            $resultados = $this->seleccionar($tabla, '*', $where, null, 1, null, 'array');
            return count($resultados) > 0 ? $resultados[0] : null;
            
        } catch (\Exception $e) {
            $this->logger->error("Error en seleccionarUno: {$e->getMessage()}");
            throw new QueryException("Error al seleccionar uno de $tabla: {$e->getMessage()}");
        }
    }
    
    /**
     * Inserta un nuevo registro
     * 
     * @param string $tabla Nombre de la tabla
     * @param array $datos Diccionario con columna => valor
     * @param bool $returnId Si true, devuelve array con info
     * @return int|array Filas afectadas o array con info
     */
    public function insertar(string $tabla, array $datos, bool $returnId = false) {
        try {
            if (empty($datos)) {
                throw new ValidationException("No se proporcionaron datos para insertar");
            }
            
            $columnas = implode(', ', array_keys($datos));
            $placeholders = implode(', ', array_fill(0, count($datos), '?'));
            $valores = array_values($datos);
            
            $query = "INSERT INTO $tabla ($columnas) VALUES ($placeholders)";
            $this->logger->query($query, $valores);
            
            list($conn, $shouldRelease, $pooledConn) = $this->getConnection();
            $resultado = $conn->executeQuery($query, $valores);
            
            // Obtener último ID insertado
            $lastId = method_exists($conn, 'getLastInsertId') ? 
                $conn->getLastInsertId() : 
                $conn->_connection->insert_id ?? null;
            
            $this->releaseConnection($pooledConn);
            
            $this->logger->info("Registro insertado en '$tabla'");
            
            if ($returnId) {
                return [
                    'rows_affected' => $resultado ?? 0,
                    'last_insert_id' => $lastId
                ];
            }
            
            return $resultado ?? 0;
            
        } catch (\Exception $e) {
            $this->logger->error("Error en insertar: {$e->getMessage()}");
            throw new QueryException("Error al insertar en $tabla: {$e->getMessage()}");
        }
    }
    
    /**
     * Inserta múltiples registros de forma optimizada
     * 
     * @param string $tabla Nombre de la tabla
     * @param array $registros Lista de arrays con los datos
     * @return int Total de filas insertadas
     */
    public function insertarMultiple(string $tabla, array $registros): int {
        try {
            if (empty($registros)) {
                return 0;
            }
            
            $total = 0;
            $this->beginTransaction();
            
            try {
                foreach ($registros as $registro) {
                    $total += $this->insertar($tabla, $registro);
                }
                $this->commit();
            } catch (\Exception $e) {
                $this->rollback();
                throw $e;
            }
            
            $this->logger->info("$total registros insertados en '$tabla'");
            return $total;
            
        } catch (\Exception $e) {
            $this->logger->error("Error en insertarMultiple: {$e->getMessage()}");
            throw new QueryException("Error al insertar múltiples en $tabla: {$e->getMessage()}");
        }
    }
    
    /**
     * Actualiza registros existentes
     * 
     * @param string $tabla Nombre de la tabla
     * @param array $datos Datos a actualizar
     * @param int|null $identificador ID del registro
     * @param array|null $where Condiciones alternativas
     * @return int Filas afectadas
     */
    public function actualizar(
        string $tabla,
        array $datos,
        ?int $identificador = null,
        ?array $where = null
    ): int {
        try {
            if (empty($datos)) {
                throw new ValidationException("No se proporcionaron datos para actualizar");
            }
            
            if ($identificador !== null) {
                $where = ['Identificador' => $identificador];
            } elseif ($where === null) {
                throw new ValidationException("Debe proporcionar identificador o where");
            }
            
            // Construir SET clause
            $setClause = implode(', ', array_map(fn($col) => "$col = ?", array_keys($datos)));
            $valores = array_values($datos);
            
            // Construir WHERE clause
            $whereConditions = [];
            foreach ($where as $col => $val) {
                $whereConditions[] = "$col = ?";
                $valores[] = $val;
            }
            
            $query = "UPDATE $tabla SET $setClause WHERE " . implode(' AND ', $whereConditions);
            $this->logger->query($query, $valores);
            
            list($conn, $shouldRelease, $pooledConn) = $this->getConnection();
            $resultado = $conn->executeQuery($query, $valores);
            $this->releaseConnection($pooledConn);
            
            $this->logger->info("$resultado registro(s) actualizado(s) en '$tabla'");
            return $resultado ?? 0;
            
        } catch (\Exception $e) {
            $this->logger->error("Error en actualizar: {$e->getMessage()}");
            throw new QueryException("Error al actualizar $tabla: {$e->getMessage()}");
        }
    }
    
    /**
     * Elimina registros de una tabla
     * 
     * @param string $tabla Nombre de la tabla
     * @param int|null $identificador ID del registro
     * @param array|null $where Condiciones alternativas
     * @return int Filas eliminadas
     */
    public function eliminar(string $tabla, ?int $identificador = null, ?array $where = null): int {
        try {
            if ($identificador !== null) {
                $where = ['Identificador' => $identificador];
            } elseif ($where === null) {
                throw new ValidationException("Debe proporcionar identificador o where");
            }
            
            // Construir WHERE clause
            $whereConditions = [];
            $valores = [];
            foreach ($where as $col => $val) {
                $whereConditions[] = "$col = ?";
                $valores[] = $val;
            }
            
            $query = "DELETE FROM $tabla WHERE " . implode(' AND ', $whereConditions);
            $this->logger->query($query, $valores);
            
            list($conn, $shouldRelease, $pooledConn) = $this->getConnection();
            $resultado = $conn->executeQuery($query, $valores);
            $this->releaseConnection($pooledConn);
            
            $this->logger->info("$resultado registro(s) eliminado(s) de '$tabla'");
            return $resultado ?? 0;
            
        } catch (\Exception $e) {
            $this->logger->error("Error en eliminar: {$e->getMessage()}");
            throw new QueryException("Error al eliminar de $tabla: {$e->getMessage()}");
        }
    }
    
    /**
     * Cuenta registros en una tabla
     * 
     * @param string $tabla Nombre de la tabla
     * @param array|null $where Condiciones opcionales
     * @return int Número de registros
     */
    public function contar(string $tabla, ?array $where = null): int {
        try {
            $query = "SELECT COUNT(*) as total FROM $tabla";
            $params = [];
            
            if ($where) {
                $conditions = [];
                foreach ($where as $col => $val) {
                    $conditions[] = "$col = ?";
                    $params[] = $val;
                }
                $query .= " WHERE " . implode(' AND ', $conditions);
            }
            
            list($conn, $shouldRelease, $pooledConn) = $this->getConnection();
            $resultado = $conn->executeQuery($query, $params);
            $this->releaseConnection($pooledConn);
            
            return $resultado[0]['total'] ?? 0;
            
        } catch (\Exception $e) {
            $this->logger->error("Error en contar: {$e->getMessage()}");
            throw new QueryException("Error al contar en $tabla: {$e->getMessage()}");
        }
    }
    
    /**
     * Verifica si existe al menos un registro
     * 
     * @param string $tabla Nombre de la tabla
     * @param array $where Condiciones de búsqueda
     * @return bool True si existe
     */
    public function existe(string $tabla, array $where): bool {
        return $this->contar($tabla, $where) > 0;
    }
    
    /**
     * Inicia una transacción
     */
    public function beginTransaction(): void {
        if ($this->inTransaction) {
            throw new TransactionException("Ya hay una transacción en curso");
        }
        
        $conn = $this->usePool ? $this->pool->acquire()->connection : $this->conexion;
        $conn->beginTransaction();
        $this->inTransaction = true;
        $this->conexion = $conn;
        $this->logger->transaction("BEGIN");
    }
    
    /**
     * Confirma una transacción
     */
    public function commit(): void {
        if (!$this->inTransaction) {
            throw new TransactionException("No hay transacción activa");
        }
        
        $this->conexion->commit();
        $this->inTransaction = false;
        $this->logger->transaction("COMMIT");
    }
    
    /**
     * Revierte una transacción
     */
    public function rollback(): void {
        if (!$this->inTransaction) {
            throw new TransactionException("No hay transacción activa");
        }
        
        $this->conexion->rollback();
        $this->inTransaction = false;
        $this->logger->transaction("ROLLBACK");
    }
    
    /**
     * Ejecuta una consulta SQL personalizada
     * 
     * @param string $query Consulta SQL
     * @param array|null $params Parámetros
     * @return mixed Resultados
     */
    public function consultaPersonalizada(string $query, ?array $params = null) {
        try {
            $this->logger->query($query, $params);
            
            list($conn, $shouldRelease, $pooledConn) = $this->getConnection();
            $resultado = $conn->executeQuery($query, $params);
            $this->releaseConnection($pooledConn);
            
            return $resultado;
            
        } catch (\Exception $e) {
            $this->logger->error("Error en consulta personalizada: {$e->getMessage()}");
            throw new QueryException("Error en consulta: {$e->getMessage()}");
        }
    }
    
    /**
     * Obtiene estadísticas del pool
     * 
     * @return array|null Estadísticas
     */
    public function getPoolStats(): ?array {
        return $this->pool ? $this->pool->getStats() : null;
    }
    
    /**
     * Cierra el pool o la conexión
     */
    public function cerrar(): void {
        try {
            if ($this->pool) {
                $this->pool->close();
                $this->logger->info("Pool de conexiones cerrado");
            } elseif ($this->conexion) {
                $this->conexion->disconnect();
                $this->logger->info("Conexión cerrada");
            }
        } catch (\Exception $e) {
            $this->logger->error("Error al cerrar: {$e->getMessage()}");
        }
    }
    
    /**
     * Destructor
     */
    public function __destruct() {
        try {
            $this->cerrar();
        } catch (\Exception $e) {
            // Ignorar errores en destructor
        }
    }
}
